nvdaHelper/remote/excel.cpp

```diff
@@ -0,0 +1,472 @@
+	long textLength=sText?sText.Length():0;
+	if(textLength==0) {
+		return 0;
+	}
+	CComPtr<IDispatch> pDispatchFont=nullptr;
+	_com_dispatch_raw_propget(pDispatchRange,XLDISPID_RANGE_FONT,VT_DISPATCH,&pDispatchFont);
+	if(!pDispatchFont) {
+		LOG_ERROR(L"range.font failed");
+		return 0;
+	}
+	double fontSize=11.0;
+	_com_dispatch_raw_propget(pDispatchFont,XLDISPID_FONT_SIZE,VT_R8,&fontSize);
+	long iFontHeight=static_cast<long>(fontSize)*-1;
+	BOOL bold=false;
+	_com_dispatch_raw_propget(pDispatchFont,XLDISPID_FONT_BOLD,VT_BOOL,&bold);
+	long iFontWeight=bold?700:400;
+	BOOL sFontItalic=false;
+	_com_dispatch_raw_propget(pDispatchFont,XLDISPID_FONT_ITALIC,VT_BOOL,&sFontItalic);
+	long sFontUnderline=0;
+	_com_dispatch_raw_propget(pDispatchFont,XLDISPID_FONT_UNDERLINE,VT_I4,&sFontUnderline);
+	BOOL sFontStrikeThrough=false;
+	_com_dispatch_raw_propget(pDispatchFont,XLDISPID_FONT_STRIKETHROUGH,VT_BOOL,&sFontStrikeThrough);
+	CComBSTR sFontName;
+	_com_dispatch_raw_propget(pDispatchFont,XLDISPID_FONT_NAME,VT_BSTR,&sFontName);
+	HDC windowDC=GetDC(hwnd);
+	HDC tempDC=CreateCompatibleDC(windowDC);
+	ReleaseDC(hwnd,windowDC);
+	HBITMAP hBmp=CreateCompatibleBitmap(tempDC,1,1);
+	HGDIOBJ hOldBmp=SelectObject(tempDC,hBmp);
+	long dpi = GetDeviceCaps(tempDC, LOGPIXELSX);
+	long iFontWidth=0;
+	long iEscapement=0;
+	long iOrientation=0;
+	long iCharSet=0;
+	long iOutputPrecision=0;
+	long iClipPrecision=0;
+	long iOutputQuality=0;
+	long iPitchAndFamily=0;
+	HFONT hFont=CreateFontW(iFontHeight, iFontWidth, iEscapement, iOrientation, iFontWeight, sFontItalic, sFontUnderline, sFontStrikeThrough, iCharSet, iOutputPrecision, iClipPrecision, iOutputQuality, iPitchAndFamily, sFontName);
+	HGDIOBJ hOldFont=SelectObject(tempDC, hFont);
+	SIZE sizl={0};
+	GetTextExtentPoint32W(tempDC, sText, textLength,&sizl);
+	SelectObject(tempDC, hOldFont);
+	DeleteObject(hFont);
+	SelectObject(tempDC, hOldBmp);
+	DeleteObject(hBmp);
+	DeleteDC(tempDC);
+	return sizl.cx;
+}
+
+__int64 getCellStates(HWND hwnd, IDispatch* pDispatchRange) {
+	__int64 states=0;
+	// If the current row is a summary row, expose the collapsed or expanded states depending on wither the inner rows are showing or not.
+	CComPtr<IDispatch> pDispatchRow=nullptr;
+	HRESULT res=_com_dispatch_raw_propget(pDispatchRange,XLDISPID_RANGE_ENTIREROW,VT_DISPATCH,&pDispatchRow);
+	if(FAILED(res)) {
+		LOG_DEBUGWARNING(L"range.entireRow failed with code "<<res);
+	}
+	if(pDispatchRow) {
+		BOOL summary=false;
+		res=_com_dispatch_raw_propget(pDispatchRow,XLDISPID_ROW_SUMMARY,VT_BOOL,&summary);
+		if(FAILED(res)) {
+			LOG_DEBUGWARNING(L"row.summary failed with code "<<res);
+		}
+		if(summary) {
+			BOOL showDetail=false;
+			res=_com_dispatch_raw_propget(pDispatchRow,XLDISPID_ROW_SHOWDETAIL,VT_BOOL,&showDetail);
+			if(FAILED(res)) {
+				LOG_DEBUGWARNING(L"row.showDetail failed with code "<<res);
+			}
+			states+=(showDetail?NVSTATE_EXPANDED:NVSTATE_COLLAPSED);
```

Comment:
Github somehow decided to discard my comment here.
Since states is supposed to be a bitmask, I'd personally prefer `states|=` over `states+=` for this and all subsequent cases.

Reply:
Yes, I think that using `+=` leads to a logic error.

Reply:
Done.